### PR TITLE
commons-mongo: refactored Filtering utility with full BsonRef and DocKey support

### PR DESCRIPTION
**Pull request by mkej**

### CODE REVIEW

File: commons-mongo/src/main/scala/com/avsystem/commons/mongo/core/ops/BaseFiltering.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import org.bson.conversions.Bson
+import org.mongodb.scala.bson.BsonValue
+
+import scala.util.matching.Regex
+
+trait BaseFiltering[T] extends Any {
+  private[core] def key: String
+  private[core] def encode(t: T): BsonValue
+
+  private def use(t: T)(f: (String, BsonValue) => Bson): Bson = {
```
Comment (Ugon) [inline commons-mongo/src/main/scala/com/avsystem/commons/mongo/core/ops/BaseFiltering.scala]: flipping those arguments would be nice for if someone wanted to use point free style xDDD

Comment (mkej) [reply commons-mongo/src/main/scala/com/avsystem/commons/mongo/core/ops/BaseFiltering.scala]: Go away.

Comment (ghik) [reply commons-mongo/src/main/scala/com/avsystem/commons/mongo/core/ops/BaseFiltering.scala]: This argument is point free.

File: commons-mongo/src/main/scala/com/avsystem/commons/mongo/core/ops/LowPrioUpdating.scala [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package com.avsystem.commons
+package mongo.core.ops
+
+import com.avsystem.commons.mongo.{BsonRef, DocKey}
+
+trait LowPrioUpdating {
+  def bsonRefUpdating[T](bsonRef: BsonRef[T]): BsonRefUpdating[T] = new BsonRefUpdating(bsonRef)
```
[the remark this answers is not in the document]
Comment (mkej) [reply commons-mongo/src/main/scala/com/avsystem/commons/mongo/core/ops/LowPrioUpdating.scala]: Yep, will add.

File: commons-mongo/src/main/scala/com/avsystem/commons/mongo/core/ops/BsonRefFiltering.scala [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package com.avsystem.commons
+package mongo.core.ops
+
+import com.avsystem.commons.mongo.BsonRef
+
+class BsonRefFiltering[T](val bsonRef: BsonRef[T])
+  extends AnyVal
```
[the remark this answers is not in the document]
Comment (mkej) [reply commons-mongo/src/main/scala/com/avsystem/commons/mongo/core/ops/BsonRefFiltering.scala]: Damn, whole base trait approach is screwed then. Any ideas how those declarations can be reused between `BsonRef` and `DocKey`?

**Pull request merged**